ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java

```diff
@@ -13,119 +13,429 @@
+        break;
+
+      case BLOB_ID_V2:
+        assertEquals("Wrong flag in blobId: " + blobId, flag, blobId.getFlag());
+        assertEquals("Wrong datacenter id in blobId: " + blobId, datacenterId, blobId.getDatacenterId());
+        assertEquals("Wrong account id in blobId: " + blobId, accountId, blobId.getAccountId());
+        assertEquals("Wrong container id in blobId: " + blobId, containerId, blobId.getContainerId());
+        break;
+    }
+  }
+
+  /**
+   * Gets the version number from a blobId string.
+   * @param blobId The blobId string to get version number.
+   * @return Version number
+   * @throws Exception Any unexpected exception.
+   */
+  private short getVersionFromBlobString(String blobId) throws Exception {
+    DataInputStream dis = new DataInputStream(new ByteBufferInputStream(ByteBuffer.wrap(Base64.decodeBase64(blobId))));
+    try {
+      return dis.readShort();
+    } finally {
+      dis.close();
+    }
+  }
+
+  /**
+   * Asserts blobId long form.
+   * @param version The version number of blobId.
+   */
+  private void assertBlobIdLongForm(short version) {
+    BlobId blobId;
+    switch (version) {
+      case BLOB_ID_V1:
+        blobId = new BlobId(referenceFlag, referenceDatacenterId, referenceAccountId, referenceContainerId,
+            referencePartitionId);
+        break;
+
+      case BLOB_ID_V2:
+        blobId = new BlobIdV2(referenceFlag, referenceDatacenterId, referenceAccountId, referenceContainerId,
+            referencePartitionId);
+        break;
+
+      default:
+        fail("Unrecognized blob version number");
+        return;
+    }
+    String blobIdStr = blobId.getID();
+    String blobIdLongForm = blobId.getLongForm();
+    String blobLongFormWithoutUuid = blobIdLongForm.substring(0, blobIdLongForm.lastIndexOf(':'));
+    StringBuilder expectedBlobIdLongFormWithoutUuidSb =
+        new StringBuilder().append("[").append(blobIdStr).append(":").append(version).append(":");
+    if (version == BLOB_ID_V2) {
+      expectedBlobIdLongFormWithoutUuidSb.append(referenceFlag)
+          .append(":")
+          .append(referenceDatacenterId)
+          .append(":")
+          .append(referenceAccountId)
+          .append(":")
+          .append(referenceContainerId)
+          .append(":");
+    }
+    String expectedBlobIdLongFormWithoutUuidStr =
+        expectedBlobIdLongFormWithoutUuidSb.append(referencePartitionId).toString();
+    assertEquals("Wrong blobId long form.", expectedBlobIdLongFormWithoutUuidStr, blobLongFormWithoutUuid);
+  }
+
+  /**
+   * A class that is used to test when {@link BlobId#CURRENT_VERSION} is set to {@link BlobId#BLOB_ID_V2}.
+   */
+  private class BlobIdV2 extends BlobId {
+    BlobIdV2(byte flag, byte datacenterId, short accountId, short containerId, PartitionId partitionId) {
+      super(flag, datacenterId, accountId, containerId, partitionId);
+    }
+
+    @Override
+    protected short getCurrentVersion() {
+      return BLOB_ID_V2;
+    }
+  }
 }
```

Comment:
since you modified the storeKey `compareTo()` and `equals()` methods and since we need to make sure that the behavior of these methods remains the same between the new code and the old code (which was used to sort and write these ids in the on-disk index segments), is it possible to add a few test cases to verify that?